Add feature-gate for SRIOV Live Migration feature

ormergi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
At the current state, if you try to migrate SRIOV VM it will fail constantly with no error from virt-api (admission or validation webhooks).
Also libvirt doesn't support migration for domain that has SRIOV host-devices attached.
We would get the following error on VMI description if we try:
```
VirtualMachineInstance migration uid bc2b1c0d-b858-44ee-b043-382caad887c8 failed. reason:Live migration failed virError(Code=84, Domain=10, Message='Operation not supported: cannot migrate a domain with <hostdev mode='subsystem' type='pci'>')
```
Even though live migration for SRIOV VMs will always fail it is not blocked on kubevirt which may cause confusion.
Thus it should be blocked and return informative error until it will be supported.
________________________________________________________

SRIOV Live Migration feature is under development,
we plan to hot-unplug SRIOV host-devices just before migration starts and hot-plug them back after the migration is finished ([design doc](https://docs.google.com/document/d/13Xne78gWheboxk_V4b5yQynYEWv5lYSTpZNICGRTeg0/edit?ts=60084ac1#heading=h.5o7sz4vqjq95))

We use VFIO driver to pass-through SRIOV host devices.
When a VFIO host-device is attached to running domain (qemu-kvm process already exists),
libvirtd adjusts the domain process resources limits.
Since virt-launcher is non-privileged, libvirtd runs as root and the domain process runs with different UID/GID,
libvirt fail to adjusts qemu-kvm process max memory limit due to the lake of CAP_SYS_RESOURCE capability [rlimits man page](https://man7.org/linux/man-pages/man3/vlimit.3.html)

________________________________________________________
This PR present new feature gate for SRIOV Live Migration
it will control when virt-launcher will run with CAP_SYS_RESOURCE capability.
This feature gate will block live migration for SRIOV VM's until it will be fully supported.

To provide better user experience It will work in a way that if SRIOVLiveMigration feature-gate is open, LiveMigration feature gate will takes affect as well.
But not the counterpart; if LiveMigration feature-gate is specified  then 
SRIOVLiveMigration will **not** take affect.
Doing so will avoid the user from enabling both `LiveMigration` `SRIOVLiveMigration` feature-gates.

This feature-gate could probably dropped once this is resolved
https://bugzilla.redhat.com/show_bug.cgi?id=1916346
    
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This PR is needed for handling SRIOV Live Migration on the target side.
Related to https://github.com/kubevirt/kubevirt/pull/4741

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add new feature-gate SRIOVLiveMigration,
controls running virt-launcher with CAP_SYS_RESOURCE.
Live migration for SRIOV VM's is blocked until fully supported.
```

ormergi commented:
/cc @EdDev @vladikr

EdDev commented on pkg/virt-config/config_test.go:
```diff
@@ -482,4 +482,38 @@ var _ = Describe("ConfigMap", func() {
 		Expect(emulation).To(BeFalse())
 	})
+
+	It("when feature-gate SRIOVLiveMigration is on, LiveMigrationEnabled should return true", func() {
```
I think this fits a table test.
You pass the feature-gates in with the expected results of the two.

This seems a classic case where it fits.

ormergi replied:
done

EdDev commented on pkg/virt-config/config_test.go:
```diff
@@ -482,4 +482,38 @@ var _ = Describe("ConfigMap", func() {
+			Data: map[string]string{virtconfig.FeatureGatesKey: "SRIOVLiveMigration"},
+		})
+		Expect(clusterConfig.SRIOVLiveMigrationEnabled()).To(BeTrue())
+
```
Newline should probably be there between the setup of the test (creating the fake) and the asserts. But I see no reason for it between the asserts.

ormergi replied:
I changed those tests to DescribeTable

davidvossel commented on pkg/virt-controller/services/template.go:
```diff
@@ -1326,21 +1327,28 @@ func (t *templateService) RenderHotplugAttachmentPodTemplate(volume *v1.Volume,
-	return res
+
+	// add CAP_SYS_RESOUCE capacility to allow Live Migration for VM with SRIOV interfaces
+	if config.SRIOVLiveMigrationEnabled() && util.IsSRIOVVmi(vmi) {
```
do we have any concern that someone could attempt to start the VMI with SRIOV, then flip this feature gate on afterwards? When that occurs, the VMI that launched before the feature gate is set will not be live migratable.

I think your logic as it is now is the right thing to do. Maybe worth simply documenting this edge case so people are aware.

EdDev replied:
I think it will actually be migratable, as this affects the target node (when the pod there is created, it should have the right capability in order to allow the hot-plug).

There is some corner cases where this will cause some issues: One of the stages is to try and revert the detached devices on the source in case the migration fails (i.e. attach them back to the domain). That one will not work if I follow your scenario correctly.

ormergi replied:
In more details , currently switching the feature-gate on affects the target pod creation stage of the migration, where it calls [`renderLaunchManifest`](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-controller/services/template.go#L318) that conditionally add the required capability according to the feature-gate state (the mentioned line).
If you have running VM with SRIOV  you want to migrate when the feature-gate is on, the target VM pod will have the required capability. 
So after the migration is finished attaching the SRIOV devices at the target should not fail.

It was also suggested to block migration for VM's with SRIOV by default if the feature gate is off
https://github.com/kubevirt/kubevirt/pull/4874#pullrequestreview-575894290

davidvossel reviewed (commented):
do we have plans for adding a functional test?

EdDev commented on pkg/virt-controller/services/template.go:
```diff
@@ -1326,21 +1327,28 @@ func (t *templateService) RenderHotplugAttachmentPodTemplate(volume *v1.Volume,
-func getRequiredCapabilities(vmi *v1.VirtualMachineInstance) []k8sv1.Capability {
-	res := []k8sv1.Capability{}
+func getRequiredCapabilities(vmi *v1.VirtualMachineInstance, config *virtconfig.ClusterConfig) []k8sv1.Capability {
+	capabilities := []k8sv1.Capability{}
```
Please do this change in a separate commit, it is confusing to see renaming changes with the other changes.

ormergi replied:
done

EdDev commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1916,47 +1917,42 @@ var _ = Describe("Template", func() {
 
 		Context("with sriov interface", func() {
+			getVMIWithSriovInterface := func(name, uid string) *v1.VirtualMachineInstance {
+				return &v1.VirtualMachineInstance{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: name, Namespace: "default", UID: types.UID(uid),
+					},
+					Spec: v1.VirtualMachineInstanceSpec{
+						Domain: v1.DomainSpec{
+							Devices: v1.Devices{
+								DisableHotplug: true,
+								Interfaces: []v1.Interface{
+									v1.Interface{
+										Name: "sriov-nic",
+										InterfaceBindingMethod: v1.InterfaceBindingMethod{
+											SRIOV: &v1.InterfaceSRIOV{},
+										},
+									},
+								},
+							},
+						},
+					},
+				}
+			}
```
It is too hard to read this now.

EdDev replied:
In a different commit I would also try to remove `DisableHotplug: true`, I do not understand how it is relevant.

ormergi replied:
I changed this function so it will be easy to understand
Also removed DisableHotplug on different commit

EdDev commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1916,47 +1917,42 @@ var _ = Describe("Template", func() {
 		})
 
 		Context("with sriov interface", func() {
+			getVMIWithSriovInterface := func(name, uid string) *v1.VirtualMachineInstance {
```
Please do not use a lambda is it is not needed. It makes it harder to figure out if an outer variable was used. Try to make it a regular function.

Regarding the name, the "get" is always overloaded. In this case `new` seems to fit better.

ormergi replied:
done

EdDev commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1951,6 +1953,20 @@ var _ = Describe("Template", func() {
 				Expect(len(pod.Spec.Containers)).To(Equal(1))
 				Expect(*pod.Spec.Containers[0].SecurityContext.Privileged).To(BeFalse())
 			})
+			It("should runs with CAP_SYS_RESOUCE capability when SRIOVLiveMigration feature-gate is on", func() {
```
nit: "runs" => "run"

ormergi replied:
Done

EdDev reviewed (changes requested):
Thanks.
Except of the inline comments, please try to add the unit tests with the production changes in the same commit.
One of the changes includes a refactoring step, please add it before you change the production code + unit tests for it.

vladikr reviewed (commented):
Here are a few thoughts I have on this topic.

Live migration with an SR-IOV card, or any host device, doesn't provide an uninterrupted migration. Therefore, it must require a user (cloud admin) to opt-in - in other words, a feature-gate. 
The currently proposed feature gate only controls the addition of a cap_sys_resource, while from my point of view we should control the process as a whole.

As I see it, we should: 
 - block live migration for VMIs with host devices. Making it by default non-migratable. See:  https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L968
Actually, #4323 makes this code nicer, but not merged yet 
- Change this code to enable live migration with SRIOV (perhaps consider all host devices, not just sriov)
 
Overall, I think an addition of such a feature gate should be done as part of a PR that enables live migration with sriov/hostdev feature.

EdDev commented:
> Here are a few thoughts I have on this topic.
> 
> Live migration with an SR-IOV card, or any host device, doesn't provide an uninterrupted migration. Therefore, it must require a user (cloud admin) to opt-in - in other words, a feature-gate.

Well, we do not provide uninterrupted migration for the network part for the primary network (pod network) while on masquerade binding. And we do not feature-gate it.
Generally speaking, I think users expect this feature to exist by default.

> The currently proposed feature gate only controls the addition of a cap_sys_resource, while from my point of view we should control the process as a whole.

As we currently need this feature-gate anyway, it seems reasonable to me.
I am not sure if it will be correct once the capability limitation will be gone, but there is no point entering into that prematurely.

> As I see it, we should:
> 
>     * block live migration for VMIs with host devices. Making it by default non-migratable. See:  https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L968
>       Actually, #4323 makes this code nicer, but not merged yet
> 
>     * Change this code to enable live migration with SRIOV (perhaps consider all host devices, not just sriov)
> 
> 
> Overall, I think an addition of such a feature gate should be done as part of a PR that enables live migration with sriov/hostdev feature.

I'm good with adding to this PR also a blocking of the migration process for SR-IOV based VMI/s.
Generalizing this for all hostdev is premature IMO, especially when there is no implementation to back that up at the moment.

We will not be able to introduce all SR-IOV live migration in one PR, therefore, the [proposal](https://docs.google.com/document/d/13Xne78gWheboxk_V4b5yQynYEWv5lYSTpZNICGRTeg0) has suggested phasing this in a way it can be taken in without causing any degradation or risk compared to the current state.
We should add this feature-gating SR-IOV migrations as a phase.

ormergi commented:
Addressed almost all comments
Changes: 
- At virt-controller block `SRIOVLiveMigration` if feature-gate is closed 
  on `VirtualMachineInstanceMigration` creation + unit tests
- Exported refactoring to commits.
- Squash unit tests and production code change commits  
- Rearranged commits refactor commits before production code change + unit tests commits
- virt-config unit tests changes tests to use `DescribeTable` 
- virt-controller template unit tests 
  - typos
  - changed removed `DisablePlugin` property
  - improved exported function readability

EdDev commented on pkg/virt-config/config_test.go:
```diff
@@ -482,4 +482,18 @@ var _ = Describe("ConfigMap", func() {
+	},
+		table.Entry("LiveMigration and SRIOVLiveMigration are closed, both should be closed", "", false, false),
+		table.Entry("LiveMigration and SRIOVLiveMigration are open, both should be open", "LiveMigration,SRIOVLiveMigration", true, true),
+		table.Entry("SRIOVLiveMigration is open, LiveMigration should be open", "SRIOVLiveMigration", true, true),
```
I was thinking about this for some time, but I have hard time to decide if it makes sense or not.
LiveMigration controls the operation for all cases, be it SR-IOV or anything else. The new one is more specific, only for SR-IOV, which makes it a but off that a specific one will control all implicitly.

@vladikr , what do you think about this case?

EdDev commented on pkg/virt-config/config_test.go:
```diff
@@ -482,4 +482,18 @@ var _ = Describe("ConfigMap", func() {
+		Expect(clusterConfig.SRIOVLiveMigrationEnabled()).To(Equal(isSRIOVLiveMigrationEnabled))
+	},
+		table.Entry("LiveMigration and SRIOVLiveMigration are closed, both should be closed", "", false, false),
+		table.Entry("LiveMigration and SRIOVLiveMigration are open, both should be open", "LiveMigration,SRIOVLiveMigration", true, true),
```
Can you use `virtconfig.LiveMigrationGate` and `virtconfig.SRIOVLiveMigrationGate` ?

ormergi replied:
Done

EdDev commented on pkg/virt-config/config_test.go:
```diff
@@ -482,4 +482,18 @@ var _ = Describe("ConfigMap", func() {
+			Data: map[string]string{virtconfig.FeatureGatesKey: openFeatureGates},
+		})
+
+		Expect(clusterConfig.LiveMigrationEnabled()).To(Equal(isLiveMigrationEnabled))
```
I could not find `isLiveMigrationEnabled` implementation, what is it?

ormergi replied:
`isLiveMigrationEnabled` and `isSRIOVLiveMigrationEnabled` are boolean that passed to the `DescribeTable` function (line 486)

EdDev reviewed (commented):
Review of "virt-config: Present new feature-gate for SRIOV live-migration" commit.

EdDev reviewed (commented):
Review commit "virt-controller, services: refactor getRequiredCapabilities":
typo in the commit message: `meaningful`.

EdDev reviewed (commented):
Review commit "virt-controller, template_test: remove DisablePlug from sriov interfa ...":

The subject line is too long, how about: `virt-controller, template: Drop DisableHotplug from SRIOV tests`

EdDev commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1964,69 +1945,45 @@ var _ = Describe("Template", func() {
-							kubev1.ResourceMemory: resource.MustParse("512Mi"),
-							kubev1.ResourceCPU:    resource.MustParse("150m"),
-						},
+				requests := v1.ResourceRequirements{
```
`requests`? Maybe you meant `requirements` ?

ormergi replied:
Yes I will change it

EdDev commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1964,69 +1945,45 @@ var _ = Describe("Template", func() {
 					},
-					Spec: v1.VirtualMachineInstanceSpec{Domain: domain},
 				}
+				domain := vmi.Spec.Domain
```
Why do you need this? It seems to me like it can be inlined.

ormergi replied:
Done

EdDev reviewed (commented):
review commit: "virt-controller, template_test: refactor sriov interface tests"

> Currently at each test of VMI with SRIOV interfaces
> rendering tests, each create a VMI object with
> all necessary modfications.

This sentence is broken and has a type. You could try to fix it a bit, but I would just drop it as the info is in the next section anyway.

EdDev commented:
Commits 2-4 of this PR is about refactoring existing test code. I would move them to a different PR to get this in faster, especially that they are valuable regardless to this PR core change.

EdDev reviewed (commented):
commit "virt-controller, services: add SYS_RESOURCE capability to virt-launcher":

The capabilities are tested at the e2e tests under "security_features_test.go", but some are specific and are checked under individual tests (e.g. ).

I think you can add this test to the SR-IOV ones: One for checking that without the feature-gate it does not have it, something that can be added to a basic test and another to a new test that does have this feature gate opened).
An example of checking the capabilities is also [here](https://github.com/kubevirt/kubevirt/blob/977c082fc22ea8d2f17fccf10897b67333efa0f1/tests/vmi_networking_test.go#L473).

EDIT: I saw you did this at the last commit.

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -1904,15 +1904,26 @@ func (d *VirtualMachineController) isPreMigrationTarget(vmi *v1.VirtualMachineIn
+	return nil
+}
+
+func (d *VirtualMachineController) checkMasquradeInterfacesForMigration(vmi *v1.VirtualMachineInstance) error {
```
I can ignore it, but this looks really ugly and messy function.

Not sure why the original and this new one is a method and not a regular function, none are using the receiver.
The name is also odd, it looks to me live a "validate" or "assert", the current name of "check" has little meaning to me.

I would just refactor this to look like this:
```
podnet := podNetwork(vmi)
if podnet := podNetwork(vmi); podnet != nil && lookupIface(podnet.Name).Masquerade == nil {
    return fmt.Errorf(...)
}
```

But if you want to leave it like this, just change the name to something I can understand from reading it. Like: `validatePodNetworkInterfaceUsesMasqueradeBinding`

ormergi replied:
I changed the function name and the code to what you suggest it does improve readability

EdDev reviewed (commented):
Review "virt-handler: export masqurade interface check for migration to function":
Note that this can be also added to the refactoring PR suggested earlier.

Edit: there is a type in the subject: masquerade

EdDev commented on tests/vmi_multus_test.go:
```diff
@@ -698,6 +698,32 @@ var _ = Describe("[Serial]SRIOV", func() {
+			pod := pods.Items[0]
+
+			foundContainer := false
+			for _, container := range pod.Spec.Containers {
+				if container.Name == "compute" {
+					foundContainer = true
+					Expect(container.SecurityContext.Capabilities.Add).
+						To((ContainElement(k8sv1.Capability("SYS_RESOURCE"))))
+					break
+				}
+			}
```
How about:
```
computeContainers := filterContainersByName("compute")
Expect(computeContainers).NotTo(BeEmpty(), ...)
for c := range computeContainers {...}
```

ormergi replied:
I found a function in tests package that does exactly that, I will use it

ormergi replied:
> For the last commit: I think you should add to `Live Migration for VM's with SRIOV interfaces should be blocked.` something like "even though the feature-gate is open".

Done

ormergi commented on tests/vmi_multus_test.go:
```diff
@@ -698,6 +698,32 @@ var _ = Describe("[Serial]SRIOV", func() {
 		}
 
+		It("should create virt-launcher pod with CAP_SYS_RESOUCE capability", func() {
+			tests.EnableFeatureGate("SRIOVLiveMigration")
```
It seems that we dont call the "big" cleaner `BeforeTestCleanup` on SRIOV tests https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_multus_test.go#L606

We have two options here, move to `BeforeTestCleanup` to `BeforeEach`
or reset kubevirt-config at the end of the tests.
Let me know what you think

EdDev replied:
In this PR scope, we should do the minimal in order not to mess up the other tests.
I think the safest path at the moment is to make sure the feature-gate is disabled when the test ends. You can use a defer in the body and call `DisableFeatureGate` or have a context that has the Enable and Disable as fixtures.

There is one disadvantage here: The log collection will not include this feature-gate, but I think this is not a problem as this is not causing a log lost.

dankenigsberg commented on pkg/virt-config/feature-gates.go:
```diff
@@ -24,19 +24,20 @@ package virtconfig
+	CPUManager             = "CPUManager"
+	IgnitionGate           = "ExperimentalIgnitionSupport"
+	LiveMigrationGate      = "LiveMigration"
+	SRIOVLiveMigrationGate = "SRIOVLiveMigration"
```
Please document the meaning (and the downsides) of your new feature somewhere.

ormergi replied:
Done

ormergi replied:
Correct me if I am wrong but currently there is no documentation that explain how a feature-gate affects the system. 
If one would like to know more then that a feature is enabled he would need to investigate the code.

I think those kind of comments should be placed right next to the deceleration (it will also appear on the IDE when you hover on it), 
until we will have a document about feature-gate with an explanation of what a feature-gate does (if there is a doc that does please let me know).

When this feature-gate will be redundant we could revert this PR (or at lease follow what it change) so this comment will also be removed

ormergi replied:
> There is documentation that handles each feature independently, e.g. https://kubevirt.io/2020/Live-migration.html
> 

I meant to a document that describes all the feature-gates we have with explanation for what they do and their side-effects
(e.g SRIOVLiveMigration side-effect is running virt-launcher with CAP_SYS_RESOURCE)

> In code we express detailed behavior using tests.
> It is fine to document what it does here, but it should not include "how" as that is an implementation detail that can change. And the how (e.g. adding the capability) can go away without removing the gating (which has other reasons to exists, like the missing devices during the migration period).
>
> So my suggestion is to leave only the first line (`// SRIOVLiveMigrationGate enable's Live Migration for VM's with SRIOV interfaces.`)
> 

In general I agree with you.
I think that SRIOVLiveMigration feature-gate is a special case because it controls when virt-launcher pod runs with extra privileges (CAP_SYS_RESOURE).
Which degrade its security and might increase attack surface.

I was thinking about leaving the first line like you suggest 
and adding a note about the capability, something like:
`When enabled virt-launcher runs with CAP_SYS_RESOURE`
Or 
`When enabled virt-launcher runs with extra privileges`
What you think?

EdDev replied:
I think you should document this in the place the action is taking place. Here you declare a flag and someone else is going to use it... the addition of the capability is there, not here.

So if you want be pedantic here, add the comment about this in the place the action is taken place, ref to a BZ and mention this is due that and can be removed when it is resolved. Finally you could also add this in the release-notes by adding it in the proper section in the PR description.

ormergi replied:
> For the last commit: I think you should add to `Live Migration for VM's with SRIOV interfaces should be blocked.` something like "even though the feature-gate is open".

Done

ormergi commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1917,6 +1917,8 @@ var _ = Describe("Template", func() {
 		})
 
 		Context("with sriov interface", func() {
+			const capSysResource = kubev1.Capability("SYS_RESOURCE")
```
Done

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -1904,18 +1904,40 @@ func (d *VirtualMachineController) isPreMigrationTarget(vmi *v1.VirtualMachineIn
 	return nil
 }
 
+func getVMIPodNetwork(vmi *v1.VirtualMachineInstance) *v1.Network {
```
`get` ==> `lookup`

I suggest you return only the name, as this is the only think you need.

ormergi replied:
Done

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -1904,18 +1904,40 @@ func (d *VirtualMachineController) isPreMigrationTarget(vmi *v1.VirtualMachineIn
+	var podNetwork *v1.Network
+	for _, network := range vmi.Spec.Networks {
+		if network.Pod != nil {
+			podNetwork = &network
```
It is unsafe to return the for-value address in general.
But you do not need all this anyway, just return the name or a the whole network as a value.
(if it still fitting to return the network itself, use the index (`return &vmi.Spec.Networks[i]`)

Note: You do not need the `podNetwork` variable as well, it is clearer that you return nil or an empty string if it was not found.

ormergi replied:
I will change it to return the name it is sufficient for now.

Out of curiosity why is it unsafe to return the for-value address?

About the note, I find function that have single return more readable and much easier to follow
what you think?

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -1938,6 +1943,16 @@ func getVMIPodNetwork(vmi *v1.VirtualMachineInstance) *v1.Network {
+func (d *VirtualMachineController) checkSRIOVInterfacesForMigration(vmi *v1.VirtualMachineInstance) error {
+	for _, iface := range vmi.Spec.Domain.Devices.Interfaces {
+		if iface.SRIOV != nil && !d.clusterConfig.SRIOVLiveMigrationEnabled() {
+			return fmt.Errorf("SRIOVLiveMigration feature-gate is closed cannot migrate VMI with SRIOV interfaces")
```
nit: After "closed" add ",".
Also, instead of "cannot" use "can not" or "can't".

ormergi replied:
Done

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -1938,6 +1943,16 @@ func getVMIPodNetwork(vmi *v1.VirtualMachineInstance) *v1.Network {
 	return podNetwork
 }
 
+func (d *VirtualMachineController) checkSRIOVInterfacesForMigration(vmi *v1.VirtualMachineInstance) error {
```
Can you use `validate` instead of `check` here as well?

ormergi replied:
Done

EdDev commented on pkg/virt-handler/vm_test.go:
```diff
@@ -1906,6 +1906,56 @@ var _ = Describe("VirtualMachineInstance", func() {
+			It("should not block migration for VMI with SRIOV interface when feature-gate SRIOVLiveMigration is on", func() {
+				vmi := v1.NewMinimalVMI("testvmi")
+				sriovInterfaceName := "sriovnet1"
+				kubevirtConfigMapFeatureGate := map[string]string{"feature-gates": "SRIOVLiveMigration"}
```
Do we have constants for these keywords? ("feature-gates", "SRIOVLiveMigration")

ormergi replied:
Nope, there is no other test in this package that handles feature-gate, you prefer to have constants for those in this file anyway?

ormergi replied:
Done

EdDev commented on tests/vmi_multus_test.go:
```diff
@@ -698,6 +698,25 @@ var _ = Describe("[Serial]SRIOV", func() {
 			}
 		}
 
+		It("[Serial]should create virt-launcher pod with CAP_SYS_RESOURCE capability", func() {
```
- The first container is marked as `[serial]` already. You do not need to mark it again here.
- Why do we need a special test for this? It will be cheaper to just add another assert to an existing one.
For example to "[test_id:1754]should create a virtual machine with sriov interface".

You can call from it to an assert helper, e.g. `assertVirtLauncherIncludesCapability(vmi, ...)

ormergi replied:
1. This test make use of `KubevirtDefaultConfig` ([comment](https://github.com/kubevirt/kubevirt/pull/4874/files/03d29a67d49a89760b6f7eb574ab3f19f0cbc9fe..f5de62693873e4ce1b483e112e71ba4a6673e1c7#r569567253)), it is expected a test to ran as serial.
I added `[serial]` for this test explicitly to make sure this test will run as serial in case it will be moved
or changed.

2. I dont think we should piggyback this test on another one, in my opinion its too cheap.
Also one could ask why do we assertVirtLauncherIncludesCapability only on the first test.
This test works and didnt took such effort anyway.
What are your concerns with having this test?

EdDev replied:
1. There are other reasons why these tests cannot run in parallel and this specific one exists in other places without such an explicit mark.
2. An e2e test is not cheap, it can take a few minutes to run the test.

EdDev commented on tests/vmi_multus_test.go:
```diff
@@ -698,6 +698,25 @@ var _ = Describe("[Serial]SRIOV", func() {
 
+		It("[Serial]should create virt-launcher pod with CAP_SYS_RESOURCE capability", func() {
+			tests.EnableFeatureGate(virtconfig.SRIOVLiveMigrationGate)
+			// Using UpdateKubeVirtConfigValueAndWait expects a test to run in serial mode
```
I do not see any other user of this function commenting about this, so there is no reason to comment on it now.
You could add a commit that comments this on the function doctext itself.

ormergi replied:
What I said wasn't accurate, from what I understnad if a test make use of `tests.KubeVirtDefaultConfig` (and not `UpdateKubeVirtConfigValueAndWait`) it is expected to run in serial mode
https://github.com/kubevirt/kubevirt/blob/master/tests/utils.go#L4291-L4300

EdDev replied:
I think there is only one kubevirt configmap per cluster, so I do not think it matters what value it gets.

EdDev replied:
Can you remove the comment?
 The commit message needs to be updated as well.

ormergi replied:
Done

EdDev commented on tests/vmi_multus_test.go:
```diff
@@ -698,6 +698,25 @@ var _ = Describe("[Serial]SRIOV", func() {
+			computeContainer := tests.GetComputeContainerOfPod(virtLauncherPod)
+			Expect(computeContainer).ToNot(BeNil(), "should get virt-launcher pod, compute container")
+			Expect(computeContainer.SecurityContext.Capabilities.Add).
+				To(ContainElement(k8sv1.Capability("SYS_RESOURCE")))
```
Please use a constant and not a string. This should be defined in the production code somewhere.

ormergi replied:
Done

ormergi replied:
Done

EdDev commented on tests/vmi_multus_test.go:
```diff
@@ -698,6 +698,25 @@ var _ = Describe("[Serial]SRIOV", func() {
+			vmi = waitVmi(vmi)
+
+			By("Looking up for VMI virt-launcher pod using VMI's label")
+			virtLauncherPod := tests.GetRunningPodByVirtualMachineInstance(vmi, tests.NamespaceTestDefault)
```
Please take the namespace from `vmi.Namespace`.

ormergi replied:
Done

ormergi replied:
Done

ormergi commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1964,69 +1945,44 @@ var _ = Describe("Template", func() {
-					Spec: v1.VirtualMachineInstanceSpec{Domain: domain1},
+				vmi := newVMIWithSriovInterface("testvmi1", "1234")
+				vmi.Spec.Domain.Resources = requirements
+				domain := vmi.Spec.Domain
```
Right I forgot to do it inline like in other test I changed

ormergi replied:
Done

ormergi commented on pkg/virt-handler/vm.go:
```diff
@@ -1904,18 +1904,40 @@ func (d *VirtualMachineController) isPreMigrationTarget(vmi *v1.VirtualMachineIn
+func validatePodNetworkInterfaceUsesMasqueradeBinding(vmi *v1.VirtualMachineInstance) error {
+	podNetwork := getVMIPodNetwork(vmi)
+
+	interfacesLookup := map[string]v1.Interface{}
```
Done

ormergi commented:
Changes: 
- Addressed code reviews comments
- Currently the code for handling Live Migration for SRIOV VM's on the source (#4741) and target nodes in not merged yet.
   In order to avoid confusion for Kubevirt users and contributors I added a commit for blocking Live Migration for VM's with
   SRIOV interfaces should be blocked.
   This commit can be removed once all the bits will get merged.

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -1904,18 +1904,38 @@ func (d *VirtualMachineController) isPreMigrationTarget(vmi *v1.VirtualMachineIn
 	}
+
+	vmiPodNetworkName := lookupVMIPodNetworkName(vmi.Spec.Networks)
+	if vmiPodNetworkName != "" && interfacesByName[vmiPodNetworkName].Masquerade == nil {
```
Self-note: I think the previous code also assumed that for each network there is an interface. So we should be safe here.

ormergi replied:
It is validated by the admmiter code, if you create an interface it must be associated with a network
This is also correct for the counterpart if you have a network it must be associated an interface.

EdDev reviewed (commented):
For the last commit: I think you should add to `Live Migration for VM's with SRIOV interfaces should be blocked.` something like "even though the feature-gate is open".

EdDev reviewed (approved):
Thank you!

EdDev commented:
//assign @AlonaKaplan

EdDev commented:
/assign @AlonaKaplan

AlonaKaplan commented:
> On SRIOV Live Migration SRIOV host-devices are un-plugged from the domain and hot-plugged after migration is finished.
> We use VFIO driver to pass-through SRIOV host devices.

Please mention it is a future plan an currently not implemented.


> This feature gate could probably dropped once this is resolved
> https://bugzilla.redhat.com/show_bug.cgi?id=1916346

And once "Implement non-root virt-launcher pods" is introduced. Right?

AlonaKaplan commented on pkg/virt-handler/vm.go:
```diff
@@ -1943,8 +1943,8 @@ func lookupVMIPodNetworkName(networks []v1.Network) string {
 	for _, iface := range vmi.Spec.Domain.Devices.Interfaces {
-		if iface.SRIOV != nil && !d.clusterConfig.SRIOVLiveMigrationEnabled() {
-			return fmt.Errorf("SRIOVLiveMigration feature-gate is closed, can't migrate VMI with SRIOV interfaces")
+		if iface.SRIOV != nil {
```
I think it is very confusing to have this `SRIOVLiveMigrationEnabled` feature gate but ignoring it.
Moreover, it is not completely ignored since it affects the `LiveMigrationEnabled` feature gate.

In this stage, I would not allow enabling this feature gate.

EdDev replied:
We need to have it in order to allow on one hand to test it opens up the capability and on the other still refusing to migrate.
The PR after this, will handle the source migration part (un-plug) where we can open this up and show through tests that migration works (with the limitation that on the target there is no plugin action done).

AlonaKaplan replied:
Let's at least have a different error message for each case.
If the `SRIOVLiveMigrationEnabled` is `false` we can keep the "virtual machine with SRIOV interfaces live migration is not supported" error message.
`SRIOVLiveMigrationEnabled` is `true`, I would add a different error message specifying the although `SRIOVLiveMigrationEnabled` is `true`, the migration is still not supported.

ormergi replied:
Done

ormergi commented:
> And once "Implement non-root virt-launcher pods" is introduced. Right?

I guess it will be introduced with running libvirtd as non root, if this is the case I am not sure how it will affect `detach-device` and `attach-device` commands.

ormergi commented:
> Please mention it is a future plan an currently not implemented.
> 
I updated the PR description

ormergi commented:
Addressed code reviews: fixed typos, and improve error messages in case feature-gate is on.
PR description is also updated

AlonaKaplan commented:
> Release note:
> Add new feature gate SRIOVLiveMigration, controls  running virt-launcher with CAP_SYS_RESOURCE

I would also mention in the release note that sriov migration is not yet supported.

EdDev commented:
> > Release note:
> > Add new feature gate SRIOVLiveMigration, controls  running virt-launcher with CAP_SYS_RESOURCE
> 
> I would also mention in the release note that sriov migration is not yet supported.

I think that the notes here are for formally declaring an external change that users should know. IMO it should only appear at the last core PR (the one that plugs the devices). Then the note will express this feature in.

But I do not feel strongly about it, whatever @AlonaKaplan prefers works with me.

ormergi commented:
> > > Release note:
> > > Add new feature gate SRIOVLiveMigration, controls  running virt-launcher with CAP_SYS_RESOURCE
> > 
> > 
> > I would also mention in the release note that sriov migration is not yet supported.
> 
> I think that the notes here are for formally declaring an external change that users should know. IMO it should only appear at the last core PR (the one that plugs the devices). Then the note will express this feature in.
> 
> But I do not feel strongly about it, whatever @AlonaKaplan prefers works with me.

We could mention that since SRIOV Live Migration is not supported yet it is blocked.

EdDev commented:
> We could mention that since SRIOV Live Migration is not supported yet it is blocked.

It makes little sense to me to say that and a week later open it. I think these changes are non-user facing, they should be hidden.

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -1943,8 +1943,11 @@ func lookupVMIPodNetworkName(networks []v1.Network) string {
-			return fmt.Errorf("SRIOVLiveMigration feature-gate is closed, can't migrate VMI with SRIOV interfaces")
+		if iface.SRIOV != nil {
+			if !d.clusterConfig.SRIOVLiveMigrationEnabled() {
+				return fmt.Errorf("SRIOV Live Migration enabled, virtual machine with SRIOV interfaces live migration is not supported yet")
```
I think they need to be formatted similarly.
How about:
`SR-IOV Live Migration enabled, however live migration of guest with SR-IOV interfaces is not supported yet`

And the second:
`SR-IOV Live Migration disabled, live migration of guest with SR-IOV interfaces is blocked`

ormergi replied:
Done

AlonaKaplan commented:
/approve

EdDev reviewed (approved):
Thank you!

ormergi commented:
@AlonaKaplan @EdDev 
Thank you!

vladikr commented on pkg/virt-handler/vm.go:
```diff
@@ -1904,15 +1904,53 @@ func (d *VirtualMachineController) isPreMigrationTarget(vmi *v1.VirtualMachineIn
+func (d *VirtualMachineController) validateSRIOVInterfacesForMigration(vmi *v1.VirtualMachineInstance) error {
+	for _, iface := range vmi.Spec.Domain.Devices.Interfaces {
+		if iface.SRIOV != nil {
+			if !d.clusterConfig.SRIOVLiveMigrationEnabled() {
+				return fmt.Errorf("SR-IOV Live Migration enabled, however live migration of guest with SR-IOV interfaces is not supported yet")
+			}
+			return fmt.Errorf("SR-IOV Live Migration disabled, live migration of guest with SR-IOV interfaces is blocked")
```
Why does it have to be so complex? We can simply say that "Live Migration of guests with SR-IOV interfaces is not supported".

AlonaKaplan replied:
I asked it so it will be clearer for a user that enabled the feature gate, that although the feature gate is enabled, the feature is not support.

vladikr replied:
I see. So this PR introduces a feature gate that doesn't enable anything :)
It's not clear to me why are doing this.

From my perspective, we should add a feature gate in a PR that introduces a feature. Otherwise, we are risking finding ourselves in an orphaned state. The Feature gate may end up being in one release and the feature in another. 

What would you think about disabling live migration with SRIOV in this PR; as this is a pressing issue?
Then add the feature gate to the PR that enables live migration with SRIOV.

EdDev replied:
> I see. So this PR introduces a feature gate that doesn't enable anything :)
> It's not clear to me why are doing this.

It does something: Adds the capability we need, blocks the SR-IOV migration and places the mechanism of the feature-gate so we can use it in the following 2 PR/s.

> From my perspective, we should add a feature gate in a PR that introduces a feature. Otherwise, we are risking finding ourselves in an orphaned state. The Feature gate may end up being in one release and the feature in another.

We have discussed this in the proposal document and some meetings.
The feature is touching a large surface and needs to be organized in a way that it can get in without causing a problem to the functionality and risk the project. Therefore we have split it into phases, this being the first.

We need the capability to continue, and using the feature-gate as a way to control it was the solution we came up with. It also fits the next phases, as we will be able to enable the migration hard block in the next phase for testing.

> What would you think about disabling live migration with SRIOV in this PR; as this is a pressing issue?
> Then add the feature gate to the PR that enables live migration with SRIOV.

This will not allow us to proceed with the planned phases and I do not see how we can push all in one change.
We are not degrading anything now, we just introduce a feature-gate that prepares the ground to control the migration and allows us at this stage to control the capability which is tested by the e2e tests.
We are not documenting this feature-gate until phase 3 (hot-plug on target) and we are not suppose to allow any misuse that will cause a problem to the whole project, at least not more than the situation today.

vladikr commented on pkg/virt-handler/vm.go:
```diff
@@ -1904,15 +1904,53 @@ func (d *VirtualMachineController) isPreMigrationTarget(vmi *v1.VirtualMachineIn
+	return ""
+}
+
+func (d *VirtualMachineController) validateSRIOVInterfacesForMigration(vmi *v1.VirtualMachineInstance) error {
```
I think you could simply use the `IsSRIOVVmi(vmi)` from pkg/util and it'll simply be 

if util.IsSRIOVVmi(vmi) && !d.clusterConfig.SRIOVLiveMigrationEnabled() {
				return fmt.Errorf("Live Migration of guests with SR-IOV interfaces is not supported")
			}

ormergi replied:
My intent was to make it easier to understand by moving the logic (that checks that the VMI has SRIOV iface and the feaute-gate is enabled) to a function that describe what it does instead of adding a comment.

vladikr reviewed (changes requested):
/hold

I apologize for being a little late to review this PR.

I have a minor request to simplify the error message. Also, please add a functional test demonstrating that we prevent migrations with SR-IOV interfaces.

AlonaKaplan commented:
Since the PR is not merged yet, please close https://github.com/kubevirt/kubevirt/pull/5016 and add it as a commit to this PR.

EdDev commented:
> /hold
> 
> I apologize for being a little late to review this PR.
> 
> I have a minor request to simplify the error message. Also, please add a functional test demonstrating that we prevent migrations with SR-IOV interfaces.

At this stage, we cannot test SR-IOV migration flow as no such cluster exists to support it (having two workers, each with its own PF on kind).
We do have it locally and the development was done with its help
Said that, we do plan now to add support for such a scenario in the kind sriov cluster, but it is planned to land only after phase 3. We will make sure to run tests with d/s setups.

The feature itself is also guarded by a feature-gate, so we should be safe here.

EdDev commented:
> Since the PR is not merged yet, please close #5016 and add it as a commit to this PR.

I do not like that we mix an openshift specific mechanism into this change. Generally speaking, we do not need to support openshift now, or in other words, we can explicitly say that we support this on openshift with a dedicated PR.

But I see that at least two are voting for this and it is a minor change, so I hope that doing this will help push this forward.

EdDev commented:
> At this stage, we cannot test SR-IOV migration flow with e2e tests as no such cluster exists to support it (having two workers, each with its own PF on kind).

I may be wrong here for this specific change. The assertion is suppose to occur before the controller will look for an available node, so we should be able to test it.
@ormergi could you please check this?

ormergi commented:
Rebased and moved #5016 change to this PR

ormergi commented:
> I may be wrong here for this specific change. The assertion is suppose to occur before the controller will look for an available node, so we should be able to test it.
> @ormergi could you please check this?

We would fail even before at VMIM creation by admission webhook

ormergi commented:
Added functest to ensure live migration is blocked for SRIOV VMI's when `SRIOVLiveMigration` feature-gate is off

ormergi commented:
> Added functest to ensure live migration is blocked for SRIOV VMI's when `SRIOVLiveMigration` feature-gate is off

Moved the test to `virt-handler: Block migration for SRIOV VM's` [commit](https://github.com/kubevirt/kubevirt/pull/4874/commits/b779846b418984f6cb7d977960d2ddf9359f22b5)

ormergi commented:
> I may be wrong here for this specific change. The assertion is suppose to occur before the controller will look for an available node, so we should be able to test it.
> @ormergi could you please check this?

@EdDev 
We would fail even before at VMIM creation by admission webhook, I added a test for blocking migration

ormergi commented:
> I may be wrong here for this specific change. The assertion is suppose to occur before the controller will look for an available node, so we should be able to test it.
> @ormergi could you please check this?

@EdDev 
We would fail even before at VMIM creation by admission webhook, I added a test for blocking migration

ormergi commented:
Changed the error message of live migration VM with SRIOV interafece

EdDev reviewed (approved):
Thank you.

@vladikr could you please have a look?

vladikr commented on tests/vmi_multus_test.go:
```diff
@@ -692,6 +693,39 @@ var _ = Describe("[Serial]SRIOV", func() {
+				To(ContainElement(k8sv1.Capability(services.CAP_SYS_RESOURCE)))
+		})
+
+		It("should block migration for SR-IOV VM's when SRIOVLiveMigration feature-gate is off", func() {
```
s/SRIOVLiveMigration/LiveMigration right?

vladikr replied:
s/VM/VMI

ormergi replied:
Done

vladikr commented:
@ormergi @EdDev @AlonaKaplan 
This looks fine to me now, considering the way the feature is going to be implemented.

Please address my minor requires to change the the text in the test name to point out that live migration from VMIs with SR-IOV interfaces is blocked. Then we can unhold it. 
Thanks!

vladikr reviewed (commented):
Thanks!
/lgtm
/hold cancel

ormergi commented:
Thanks!

ormergi commented:
/retest
